### PR TITLE
Imperative API 

### LINKED ISSUE #899: Imperative API for FunSAT
The FunSAT solver exposes a purely functional API which isn't convenient while executing a SMT-LIB script. Exposing a new imperative API simplify a lot the `Frontend` module.

The frontend `satml_frontend` exposes the same functional API built on the top of an imperative one... We may only expose an imperative API for all the SAT solvers and the current functional one of FunSAT will be an internal API only.

### PR DESCRIPTION
**Pull request by Stevendeo**
Tackles #899 by making both Sat API imperative.
- The imperative is made... well, imperative; it still relies on `Satml_frontend` for properly wrapping calls to `Satml`. We should be able to merge them, but I choosed not to in this PR to avoid spaghettification.
- The functional sat now has an imperative frontend upadting a reference to the environment.

### CODE REVIEW

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [inline src/bin/common/solving_loop.ml:144]:
```suggestion
      List.iter
        (FE.process_decl ~hook_on_status:Frontend.print_status ftdn_env)
        cnf;
```

File: src/lib/frontend/frontend.ml
Comment (bclement-ocp) [inline src/lib/frontend/frontend.ml:225]: I think this is duplicated with the info in the sat now?

File: tests/smtlib/testfile-get-info1.dolmen.expected
Comment (bclement-ocp) [inline tests/smtlib/testfile-get-info1.dolmen.expected:4]: Hm, that is surprising… This may indicate that the change is not as innocuous as we thought — unless this is the weak table somehow?

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]: Probably can remove `e` from the arguments

**Comment by Stevendeo** [comment]
There is a test that does not pass on the CI, but it works on my side...

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]:
```suggestion
            env.proxies <- proxies_mp;
```

**Comment by Stevendeo** [comment]
The test does not seem to be executed with `make runtest`, when I play it manually it does not pass on my side as well.

**Review by bclement-ocp — COMMENTED** [review]
This looks reasonable except for the change in step limit which is a bit worrying. Maybe we should run benches on ae-format just to check that everything is A-OK before merging?

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:144]: I don't like the `;` for large (i.e., 2 lines or more) chunks of code.

File: src/lib/frontend/frontend.ml
Comment (Stevendeo) [reply src/lib/frontend/frontend.ml:225]: Where is the information in the SAT? You mean by calling `get_unknown_reason`?

Comment (bclement-ocp) [reply src/lib/frontend/frontend.ml:225]: Yes. But it's orthogonal to this PR anyways.

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [reply src/bin/common/solving_loop.ml:144]: And I don't like `let () =` ever so where does that lead us :)

We don't have a style guide so I won't bikeshed, but you use the exact syntax I suggested in `worker_js.ml` so please at least be consistent.

File: src/lib/reasoners/satml_frontend.ml
Comment (bclement-ocp) [inline src/lib/reasoners/satml_frontend.ml]: Can you explain why this is moved here and why it is correct? Is it related to the test failure you observed?

Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: In the old version, if `acc` is empty, the objective update will not be taken into account because an `I_dont_know` exception is raised. It only is otherwise that the objectives must be updated.

Comment (bclement-ocp) [reply src/lib/reasoners/satml_frontend.ml]: In the old version, `env.objectives` is a reference that is shared across all versions of `env` (including the one in `unknown_exn`), so in fact the objectives should still get updated even though that is confusing control flow. Is this the change that fixes the test failure?

Comment (Stevendeo) [reply src/lib/reasoners/satml_frontend.ml]: Oh you're right I forgot it was a reference; and no, it is not what fixes the test failure.

Comment (Halbaroth) [reply src/lib/reasoners/satml_frontend.ml]: Actually the goal of this reference isn't to be able to retrieve objectives (currently, objectives are stored in a struct from `Models.ml` and the reference `last_saved_model` gives us access to them) but to keep in mind we're exploring branches beyond SAT because we're looking for a better model during the optimization. If we got an `unsat` answer  in this context, it means the last attempt to improve the model failed but the previous one is correct and we have to return `unknown` for the frontend. The critical code is here:
 ```ocaml
    | IUnsat (env, _) as e ->
      if !(env.objectives) == None then raise e;
      (* TODO: put the correct objectives *)
      raise (I_dont_know env)
```

File: src/bin/common/solving_loop.ml
Comment (Halbaroth) [reply src/bin/common/solving_loop.ml:144]: I agree that we should write style guide (not too long). I prefer using `let () =` too but it doesn't really matter.

File: src/lib/reasoners/fun_sat.mli
Comment (Halbaroth) [inline src/lib/reasoners/fun_sat.mli]: The comment is a bit confusing because we don't what `the solver` means here? Another SAT solver? I suggest:

```suggestion
(** A functional SAT solver implementation. *)
```

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:144]: > We don't have a style guide so I won't bikeshed, but you use the exact syntax I suggested in worker_js.ml so please at least be consistent.

You do realize all I want now is to update `worker_js.ml`, do you? :rofl:

Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:144]: More seriously, my first message was just explaining why I actually chose to use a `let ()` syntax, and I am sorry if you took it personally. I met people who liked `;` and others who liked `let ()`, and I can't guess what will be the opinion of my reviewer on the matter in advance.

If you want Alt-Ergo to follow a style guide for that matter, I will be happy to create it with you actually!

File: src/lib/reasoners/satml_frontend.ml
Comment (Halbaroth) [inline src/lib/reasoners/satml_frontend.ml:1271]: Probably the worst way to write for loops now :) but I agree that getting ride of this heresy is out of the scope of this PR.

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [reply src/bin/common/solving_loop.ml:144]: > You do realize all I want now is to update worker_js.ml, do you? 🤣

I would like you to use the same format in both cases but as I said: we don't have a style guide currently, and so I won't bikeshed the alternative you choose — update either `worker_js.ml` or `solving_loop.ml` I won't care.

> If you want Alt-Ergo to follow a style guide for that matter, I will be happy to create it with you actually!

Having a style good sure would be a good thing (I guess #634 should be part of that discussion as well)! The existing code is a bit of a mess and mixes different styles in somewhat arbitrary ways. Opened #967

**Comment by bclement-ocp** [comment]
> Oh you're right I forgot it was a reference; and no, it is not what fixes the test failure.

Well don't leave us hanging — what did, then?

**Review by Halbaroth — APPROVED** [review]

**Comment by Stevendeo** [comment]
> Well don't leave us hanging — what did, then?

Just a `let env =` that I forgot to replace by its proper side effect; the heuristic mechanism that forces the Normal strategy depends on a counter (`last_forced_normal`) that is reinitialized every time it triggers. The counter was not reinitialized, and the force normal behavior was looping. I am not sure of the details regarding the raised exception though, maybe there is a subtle bug hiding behind this. All I managed to find is that the culprit was the first assertion of `Satml.update_lazy_cnf`, which triggers when trying to add twice the same formula in the environment (also, there is a generic exception catcher in satml_frontend that breaks my callstack and I really want to remove it).

**Comment by bclement-ocp** [comment]
Ah, you must have changed it in a force push then, which is why I didn't see it. The question was mostly to know if it was a simple omission or if there was some trickery involved, looks like it is the first case if there was a `let env = { env with xxx } in` lying around. Please run ae-format benches then we should be good to merge!

> there is a generic exception catcher in satml_frontend that breaks my callstack and I really want to remove it

Yeah, this `assert false` is regularly annoying, we should just re-raise the exception there instead I guess, would be happy to accept a PR that does this.

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:144]: Here is a proposition: as long as there is no rule about something, the decision goes to the person who wrote the code. We can't afford to lose hours of reflexion on stuff that are not specified.

**Comment by Stevendeo** [comment]
> Yeah, this assert false is regularly annoying, we should just re-raise the exception there instead I guess, would be happy to accept a PR that does this.

Wouldn't that break the callstack as well? I would rather remove this catch (or replace it with the actual exception that are raised by the wrapped function).

**Comment by bclement-ocp** [comment]
> Wouldn't that break the callstack as well? I would rather remove this catch (or replace it with the actual exception that are raised by the wrapped function).

No, OCaml should properly preserves the call stack when re-raising exceptions.

**Comment by Stevendeo** [comment]
Now I'm thinking of it, maybe the exception was changed to an assert false is due to the fact that it may raise other exception (I_dont_know, Sat) that could be catched at a higher level. If we are sure it is not the case, why not just removing this code?

File: src/bin/common/solving_loop.ml
Comment (bclement-ocp) [reply src/bin/common/solving_loop.ml:144]: I am OK with that. I am not trying to force your hand in any way, and I have stated repeatedly that *I am OK with keeping `let () =` here*. I am simply asking that whatever style is used, the code at least be consistent within a single PR, is that really too much to ask?

**Comment by bclement-ocp** [comment]
> Now I'm thinking of it, maybe the exception was changed to an assert false is due to the fact that it may raise other exception (I_dont_know, Sat) that could be catched at a higher level. If we are sure it is not the case, why not just removing this code?

As a matter of fact I don't think we are sure that is not the case (it shouldn't be the case but… bugs exist I guess). In this case, the proper thing to do would be to raise an error (from the `Errors` module) instead (raising from within an exception handler, rather than asserting, should also preserve the call stack).

Or we could keep the `assert false` but add the backtrace to the debug message (using `Printexc.get_backtrace ()`), which maybe is the better solution.

File: src/bin/common/solving_loop.ml
Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:144]: > I am not trying to force your hand in any way... 

You are.

Comment (bclement-ocp) [reply src/bin/common/solving_loop.ml:144]: I think there is clearly a misunderstanding somewhere. I am sorry if this is the way your perceive what I am saying, because that is definitely not the intent.

Here is how the conversation went from my side:

 - I make a *suggestion* to change from `let () = ..` to using the sequence operator
 - You refuse to do it because you do not like the sequence operator for long expressions. That is fine!
 - I *accept that*, but point out that you did use the sequence operator in an exact copy of the same code, and ask you to be consistent (which is more important than any specific style being used)
 - You say "now I just want to update the other file!"
 - *I am OK with that*, which I state. I thought that we agreed, then, and that it was the end of it
 - The discussion switches to "what should we do when something is not specified" and to let the author make decisions
 - I agree with that (most of the time, style suggestions are that, *suggestions*), but there is a change that I thought we agreed on making (see above) and that was not done, which I point out, a bit sharply I admit, I am sorry for that.

To reiterate:
 - I have no strong feelings on the specific style being used
 - I do believe that it is important to be consistent to improve maintainability and clarity of the code
 - I want the PR to move forward based on a mutual understanding between author and reviewer
 - I am not going to hold up the PR for this (the only reason I have not approved yet is that I am waiting for the bench results)
 - If you prefer not to make any change here, please let me know (I guess that is the case, but again — this is not what I initially understood from the conversation); I am happy to move the PR forward based on mutual understanding.

I hope this helps clarify things, happy to have a chat with you one-on-one to clear things up if needed.

I understand that style preferences are subjective, and I respect your perspective on the matter. Looking forward to working with you on a style guide that can streamline these issues in the future :)

Comment (Stevendeo) [reply src/bin/common/solving_loop.ml:144]: Okay, I think I see where the misunderstanding is.

> You do realize all I want now is to update worker_js.ml, do you? 

This was a joke (which is what I meant in my next message starting with 'More seriously, ...') based on the fact that I believe the choice of `let ()` and `;` syntax can coexist without reducing readibility and I was making fun of my contradiction spirit :wink: I did not plan on updating `worker_js.ml` as I said, for that would be very aggressive way to express my stubbornness.

This conversation went a little too far :sweat_smile: sorry about that!

Comment (bclement-ocp) [reply src/bin/common/solving_loop.ml:144]: Ah, I understand better now, jokes are not always easy to tell apart in text format! Glad we could clear that up.

> I did not plan on updating worker_js.ml as I said, for that would be very aggressive way to express my stubbornness.

I don't think it would have been aggressive, given that it was one of the two alternatives I was initially asking for.

**Comment by bclement-ocp** [comment]
Benches look good — let's merge. Thanks!

**Pull request merged**
